### PR TITLE
Add an in-memory log provider to the test server

### LINKED ISSUE #93: Consider a "testing logger" as part of the testing framework
It would be nice to be able to test log entries easily.

Additionally, it would be nice to allow users to register their own logging factories for DI. (This may already be possible.)

### PR DESCRIPTION
**Pull request by jskeet**
(The docs state some limitations at the moment, but I think it's worth including even so.)

Fixes #93.

### CODE REVIEW

File: examples/Google.Cloud.Functions.Examples.IntegrationTests/SimpleDependencyInjectionTest.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+namespace Google.Cloud.Functions.Examples.IntegrationTests
+{
+    public class SimpleDependencyInjectionTest
+    {
+        [Fact]
+        public async Task LogEntryIsRecorded()
+        {
+            using (var server = new FunctionTestServer<SimpleDependencyInjection.Function>())
+            {
+                // We shouldn't have any log entries at the start of the test.
```
Comment (amanda-tarafa) [inline examples/Google.Cloud.Functions.Examples.IntegrationTests/SimpleDependencyInjectionTest.cs:29]: Maybe the functions framework itself could, at some point, be configurable to log startup?

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs [hunk 2]
```diff
@@ -65,6 +70,28 @@ public FunctionTestServer(Type functionType)
         /// <returns>An <see cref="HttpClient"/> that will connect to <see cref="Server"/> by default.</returns>
         public HttpClient CreateClient() => Server.CreateClient();
 
+        /// <summary>
+        /// Clears all logs.
+        /// </summary>
+        public void ClearLogs() => _testLoggerProvider.Clear();
+
+        /// <summary>
+        /// Returns a list of log entries for the given category name. If no logs have been
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:79]: nit: for the given type?

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs [hunk 3]
```diff
@@ -65,6 +70,28 @@ public FunctionTestServer(Type functionType)
...
+        /// written for the given category, an empty list is returned.
+        /// </summary>
+        /// <param name="type">The type from which to derive the category name.</param>
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:82]: nit: derive -> obtain or build?

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs [hunk 4]
```diff
@@ -65,6 +70,28 @@ public FunctionTestServer(Type functionType)
...
+        /// <returns>A list of log entries for the given category name.</returns>
+        public List<TestLogEntry> GetLogEntries(Type type) => GetLogEntries(type.FullName ?? "");
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:84]:
```suggestion
        public List<TestLogEntry> GetLogEntries(Type type) => GetLogEntries(type?.FullName ?? "");
```
(I hope ?? doesn't "translate" NullReferenceException to null :))

File: examples/Google.Cloud.Functions.Examples.IntegrationTests/SimpleDependencyInjectionTest.cs (hunk 1 above)
Comment (jskeet) [reply examples/Google.Cloud.Functions.Examples.IntegrationTests/SimpleDependencyInjectionTest.cs:29]: It does log startup, but that won't log anything into the function's category. Will clarify this comment.

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs (hunk 2 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:79]: Done.

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs (hunk 3 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:82]: I think derive is *correct*, but I don't want folks to think it's got anything to do with inheritance. Have reworded - it's a bit ugly, but it'll do.

File: src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs (hunk 4 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Invoker.Testing/FunctionTestServer.cs:84]: Have made it a precondition instead - and below for the category name.

**Pull request merged**